lib/rpmsg/rpmsg_virtio.c

```diff
@@ -102,14 +103,16 @@ static void rpmsg_virtio_return_buffer(struct rpmsg_virtio_device *rvdev,
 		/* Initialize buffer node */
 		vqbuf.buf = buffer;
 		vqbuf.len = len;
-		virtqueue_add_buffer(rvdev->rvq, &vqbuf, 0, 1, buffer);
+		ret = virtqueue_add_buffer(rvdev->rvq, &vqbuf, 0, 1, buffer);
+		RPMSG_ASSERT(ret == VQUEUE_SUCCESS, "add buffer failed\r\n");
 	}
 #endif /*VIRTIO_DEVICE_ONLY*/
 
 #ifndef VIRTIO_DRIVER_ONLY
 	if (role == RPMSG_REMOTE) {
 		(void)buffer;
-		virtqueue_add_consumed_buffer(rvdev->rvq, idx, len);
+		ret = virtqueue_add_consumed_buffer(rvdev->rvq, idx, len);
+		RPMSG_ASSERT(ret == VQUEUE_SUCCESS, "add consumed buffer failed\r\n");
```

Comment:
Can we print value of ret right before RPMSG_ASSERT using metal_err() macro? This way we know error code from add_buffer and add_consumed_buffer.

Reply:
Nice catch, how about adding a new marco `RPMSG_ASSERT_WITH_RET()` like this:
```c
#define RPMSG_ASSERT_WITH_RET(_exp, _msg, _ret) do { \
		if (!(_exp)) { \
			metal_log(METAL_LOG_EMERGENCY, \
				  "FATAL: %s, ret=%d - "_msg, __func__, ret); \
			metal_assert(_exp); \
		} \
	} while (0)
```
Because I noticed that other places use `RPMSG_ASSERT` also do not print the return value,  we can replace these `RPMSG_ASSERT` to `RPMSG_ASSERT_WITH_RET`.

Reply:
Do we really need to add extra code on something that should not happen and seems result from a memory corruption?
In my opinion the `RPMSG_ASSERT` should be sufficient

Reply:
@tnmysh @arnopo Both are good for me. 
In addition, even no memory corruption, as I mentioned in https://github.com/OpenAMP/open-amp/pull/524, return a same buffer to the virtqueue twice will lead this error too, so I want to add this assert to check other possible issues.

Reply:
@arnopo Ok, in that case I am okay with only RPMSG_ASSERT. Yes we don't need extra #define.

Reply:
Is it this patch ready for merge? @arnopo .